File: genetic-algorithm-c/src/member.c
#include <string.h>
#include <stdlib.h>
#include <math.h>

#include "member.h"

static const char target[] = "I'm Skynet and I will be back!";

static char random_char() {
    return rand()%('{' - ' ') + ' ';
}

void member_random(void *member) {
    const size_t len = strlen(target);

    for(size_t i=0; i<len; i++) {
        ((member_t *)member)->content[i] = random_char();
    }
}

void member_crossover(void *member, const void *parent1, const void *parent2) {
    const size_t len = strlen(target);
    const size_t midpoint = rand()%len;

    for(size_t i=0; i<midpoint; i++) {
        ((member_t *)member)->content[i] = ((member_t *)parent1)->content[i];
    }
    for(size_t i=midpoint; i<len; i++) {
        ((member_t *)member)->content[i] = ((member_t *)parent2)->content[i];
    }

    if((rand()%100)<5) {
        const size_t index = rand()%len;
        const char character = random_char();

        ((member_t *)member)->content[index] = character;
    }
}

float member_fitness(const void *member) {
    const size_t len = strlen(target);
    size_t diff = 0;

    for(size_t i=0; i<len; i++) {
        if(((member_t *)member)->content[i]!=target[i]) {
            diff++;
        }
    }

    if(diff==0) {
        return INFINITY;
    }

    return 1.f/diff;
}

File: rust/genetic-algorithm/src/member.rs
use std::fmt;
use rand::prelude::*;
use super::genetic::MemberInterface;

#[derive(Clone)]
pub struct Member {
    content: String,
}

impl Member {
    const TARGET: &'static str = "I'm Skynet and I will be back!";
}

impl MemberInterface for Member {
    fn random(rng: &mut ThreadRng) -> Self {
        let random_string: String = (0..Self::TARGET.len())
            .map(|_| rng.gen_range(b' '..b'}' + 1) as char)
            .collect();

        Self {
            content: random_string
        }
    }

    fn crossover(rng: &mut ThreadRng, parent1: &Member, parent2: &Member) -> Self {
        let mut child: String = String::new();

        for (ch1, ch2) in parent1.content.chars().zip(parent2.content.chars()) {
            let random = rng.gen::<bool>();
            let ch = if random { ch1 } else { ch2 };

            child.push(ch);
        }

        let random = rng.gen::<f32>();
        if random<0.05_f32 {
            let mut chars: Vec<char> = child.chars().collect();
            let index = rng.gen_range(0..chars.len());
            let character = rng.gen_range(b' '..b'}' + 1) as char;
            chars[index] = character;
            child = chars.into_iter().collect();
        }

        Self {
            content: child
        }
    }

    fn fitness(&self) -> f32 {
        let diff = self
            .content
            .chars()
            .zip(Self::TARGET.chars())
            .filter(|&(char1, char2)| char1 != char2)
            .count();

        if diff==0 {
            //return f32::INFINITY;
            return f32::MAX;
        }

        return 1_f32 / diff as f32;
    }
}

impl fmt::Display for Member {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.content)
    }
}
